File: spring/src/main/java/io/github/wavesync/repository/UserResonatorRepository.java
// MIGRATED to fastapi/app/repositories/user_resonator_repository.py
package io.github.wavesync.repository;
import io.github.wavesync.entity.UserResonator;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;


@Repository
public interface UserResonatorRepository extends JpaRepository<UserResonator, Long> {

    @Modifying
    @Query("""
        select ur.id
        from UserResonator ur
        where ur.resonatorMaster.name = :name
          and ur.isDeleted = false
    """)
    List<Long> findIdsByResonatorName(@Param("name") String name);


    @Modifying
    @Query("""
        update UserResonator ur
        set ur.isDeleted = true
        where ur.id in :ids
          and ur.isDeleted = false
    """)
    void softDeleteByIds(@Param("ids") List<Long> ids);


    @NotNull
    @Query("""
        select ur
        from UserResonator ur
        join fetch ur.resonatorMaster
        join fetch ur.weaponMaster
        join fetch ur.finalStat
        where ur.id = :userResonatorId
          and ur.isDeleted = false
    """)
    Optional<UserResonator> findById(@NotNull Long userResonatorId);

    @Query("SELECT DISTINCT ur FROM UserResonator ur " +
            "JOIN FETCH ur.resonatorMaster " +
            "JOIN FETCH ur.weaponMaster " +
            "JOIN FETCH ur.finalStat " +
            "LEFT JOIN FETCH ur.userEchoes " +
            "WHERE ur.id = :userResonatorId AND ur.isDeleted = false")
    Optional<UserResonator> findByIdForUpdate(Long userResonatorId);
}

File: fastapi/app/repositories/user_resonator_repository.py
from typing import List, Optional

from sqlalchemy import select, update
from sqlalchemy.orm import Session, joinedload, selectinload

from app.models.resonator_master import ResonatorMaster
from app.models.user_echo import UserEcho
from app.models.user_resonator import UserResonator


def save(db: Session, user_resonator: UserResonator) -> UserResonator:
    # 커밋은 호출부 책임. relationship 자식은 기본 save-update cascade로 같은 flush에 반영된다.
    db.add(user_resonator)
    return user_resonator


def find_ids_by_resonator_name(db: Session, name: str) -> List[int]:
    stmt = (
        select(UserResonator.id)
        .join(ResonatorMaster, UserResonator.resonator_master_id == ResonatorMaster.id)
        .where(ResonatorMaster.name == name, UserResonator.is_deleted.is_(False))
    )
    return list(db.scalars(stmt))


def soft_delete_by_ids(db: Session, ids: List[int]) -> None:
    stmt = (
        update(UserResonator)
        .where(UserResonator.id.in_(ids), UserResonator.is_deleted.is_(False))
        .values(is_deleted=True)
    )
    db.execute(stmt)
    # 커밋은 호출부 책임


def find_by_id(db: Session, user_resonator_id: int) -> Optional[UserResonator]:
    stmt = (
        select(UserResonator)
        .options(
            joinedload(UserResonator.resonator_master),
            joinedload(UserResonator.weapon_master),
            joinedload(UserResonator.final_stat),
        )
        .where(UserResonator.id == user_resonator_id, UserResonator.is_deleted.is_(False))
    )
    return db.scalar(stmt)


def find_by_id_for_update(db: Session, user_resonator_id: int) -> Optional[UserResonator]:
    stmt = (
        select(UserResonator)
        .options(
            joinedload(UserResonator.resonator_master),
            joinedload(UserResonator.weapon_master),
            joinedload(UserResonator.final_stat),
            # 컬렉션은 joinedload 대신 selectinload로 배치 조회 (부모 행 중복 방지, N+1 방지).
            # user_echo_subs까지 순회하므로 한 단계 더 체이닝한다.
            selectinload(UserResonator.user_echoes).selectinload(UserEcho.user_echo_subs),
        )
        .where(UserResonator.id == user_resonator_id, UserResonator.is_deleted.is_(False))
    )
    return db.scalar(stmt)
